File: Pr12/prog.c
#include <stdio.h>
#include <stdlib.h>
/*______________________________________________________________________________
______________________________________________________________________________*/

int main(void)
{
	void LeerVector(int, char *, double *);
	double ProductoEscalar(int, double *, double *);
	void FError(FILE* , char *, int);

	int    n;
	double pe;

	double *v1;
        double *v2;

	LeeNComponentes:
		printf("\n Introduce el numero de componentes: ");
		scanf("%d", &n);
	if (n < 1) goto LeeNComponentes;

        // Memory allocation
        v1 = (double *)malloc(sizeof(double) * n);
        if(v1==NULL) FError(stderr,"problema al crear el puntero 1",2);
        v2 = (double *)malloc(sizeof(double) * n);
        if(v2==NULL) FError(stderr,"problema al crear el puntero 2",2);
        
	printf("\n Introduzca vector v1[]:\n\n");
	LeerVector(n, "v1", v1);

	printf("\n Introduzca vector v2[]:\n\n");
	LeerVector(n, "v2", v2);

	pe = ProductoEscalar(n, v1, v2);

	printf("\n Producto escalar de v1[] * v2[] = %25.15le\n\n", pe);  

	fflush(stdin); getchar();
	exit(0);
}


/*________________________________________________________________LeerVector__*/

void LeerVector(int n, char *nombre, double *v)
{
	int i;

	for (i = 0; i < n; i++) {
		printf("       %s [%6d] = ", nombre, i);
		scanf("%le", &v[i]);
	}
}

/*_____________________________________________________________ProductoEscalar*/
double ProductoEscalar(int n, double *v1, double *v2){
    double pe;
    
    for(; n>=0; n--) 
        pe += v1[n]*v2[n];
    return pe;
}

/*____________________________________________________________________FError__*/

void FError(FILE* fperr, char *error, int ierr)
{
	fprintf(fperr, "\n prueba:: Error: %s\n", error);

	fprintf(stderr, "\n Fin de Ejecucion. Oprima RETURN para salir.");
	fflush(stdin); getchar();
	exit(ierr);
}

File: Pr12/prog.f
c_______________________________________________________________________________
c_______________________________________________________________________________

      implicit real*8 (a-h,o-z)

      allocatable v1(:), v2(:)

    1 write(6,100) ' Introduce el numero de componentes: '
  100   format(/a,$)
        read(5,*) n
      if (n.lt.1) goto 1

      allocate (v1(n), STAT=ist)
      if(ist.ne.0)
     +call FError(0,'fallo indeterminado con el vector 1',2)
      allocate (v2(n), STAT=ist)
      if(ist.ne.0)
     +call FError(0,'fallo indeterminado con el vector 1',2)
      
      
      write(6,'(/a/)') ' Introduzca vector v1():'
      call LeerVector(n,'v1',v1)

      write(6,'(/a/)') ' Introduzca vector v2():'
      call LeerVector(n,'v2',v2)

      pe = ProductoEscalar(n, v1, v2);

      write(6,200) pe
  200 format(/' Producto escalar de v1() * v2() = ',d25.16/)  

      read(5,*)
      call exit(0);

      end


c_____________________________________________________________________LeerVector
      Subroutine LeerVector(n,nombre,v)
      implicit real*8 (a-h,o-z)

      character*(*) nombre
      dimension v(n)

      do i=1,n
        write(6,100) nombre, i
  100   format('       ',a,' (',i6,') = ',$)
        read(5,*) v(i)
      enddo

      return
      end
c________________________________________________________________ProductoEscalar
      function ProductoEscalar(n,v1,v2)
      implicit real*8 (a-h,o-z)

      dimension v1(n), v2(n)
      pe = 0.0d+00

      do i=1,n
         pe = pe + v1(i)*v2(i)
      enddo

      ProductoEscalar = pe
      end
c_________________________________________________________________________FError
      Subroutine FError(logerr,error,ierr)
      character*(*) error

      write(logerr,100) error
  100 format(/,' prueba:: Error: ',a)

      write(logerr,'(/a,$)')
     .      ' Fin de Ejecucion. Oprima RETURN para salir.'
      read(5,'()')
      call exit(ierr)

      return
      end
